storage/docs/rest.md

```diff
@@ -1,69 +1,117 @@
 ETag: <etag>
 ```
 
-## Get a file
+When uploading for the first time, the ```If-Match``` header should be set to '*'. If the ```If-Match``` fails,
+then a 412 (Precondition Failed) should be returned.
+
+The ```Libero-file-id``` value is a reference to the storage entity's internal id for a file. When uploading to the
+same URL, a new value for ```Libero-file-id``` should be returned.
+
+## Retrieving a file
+
+To retrieve a file in the namespace ```libero``` and with key of ```directory/file.pdf```:
 
 ```
-GET /files/namespace/directory/file.ext
-#Prefer: original
+GET /files/libero/directory/file.pdf
 
 Response:
 Code: 200
 Content-Type: application/pdf
 Content-Length: 12345
-Libero-file-tags: "key=value,key2=value2,filename=bar.pdf"
-Link: "<http://sharedlink>;rel=????"
-Link: "<http://publiclink>;rel=????"
+Libero-file-tags: "filename=original-file.pdf,key1=value=1"
+Libero-file-id: bd1c9a15-bc18-4151-a4eb-9f45cedb9700
+Link: "<http://storage-url/path/to/file>;rel=public"
 Last-Modified: 2019-08-19T14:40:04.123456
 ETag: <etag>
+
 FILE DATA
 ```
 
-## Get file meta data
+The returned Link header value is the public url if the underlying storage is public.
+
+## Retrieving a file meta data
+
+### Public storage
+
+To retrieve the metadata of a file in the namespace ```libero``` and with key of ```directory/file.pdf```:
 
 ```
-HEAD /files/namespace/directory/file.ext
+HEAD /files/libero/directory/file.pdf
+```
 
-Response:
+The response should be:
+
+```
 Code: 200
 Content-Type: application/pdf
 Content-Length: 12345
-Libero-file-tags: "key=value,key2=value2,filename=bar.pdf"
-Link: "<http://sharedlink>;rel=????"
-Link: "<http://publiclink>;rel=????"
+Libero-file-tags: "filename=original-file.pdf,key1=value=1"
+Libero-file-id: bd1c9a15-bc18-4151-a4eb-9f45cedb9700
+Link: "<http://storage-url/path/to/file>;rel=public"
 Last-Modified: 2019-08-19T14:40:04.123456
 ETag: <etag>
 ```
 
-## Get file directly
-> What is this?
+The Link value is a direct url to the stored file.
+
+### Private storage
+
+If the underlying storage is private, then the following request should be used
 
 ```
-GET http://url/foo/bar.pdf
+HEAD /files/libero/directory/file.pdf
+Libero-url-type: signed
+```
 
-Response:
+The following response should be returned
+```
 Code: 200
 Content-Type: application/pdf
 Content-Length: 12345
-Libero-file-tags: "key=value,key2=value2,filename=bar.pdf"
-Link: "<http://sharedlink>;rel=????"
-Link: "<http://publiclink>;rel=????"
+Libero-file-tags: "filename=original-file.pdf,key1=value=1"
+Libero-file-id: bd1c9a15-bc18-4151-a4eb-9f45cedb9700
+Link: "<http://storage-url/path/to/file?token=abcdef>;rel=signed"
 Last-Modified: 2019-08-19T14:40:04.123456
 ETag: <etag>
-
-FILE DATA
 ```
+
+The Link value should be a signed url that can subsquently be used to access the file directly from
+the private storage entity.
+
+Note that if the ```Libero-url-type``` header is not sent then no Link header should be returned in the response.
```

Comment:
Definitely needs a `Vary`; pretty custom though. If it's _really_ needed (ie what's the problem with always returning it if you have access?), then a [`Prefer` header](https://tools.ietf.org/html/rfc7240) is preferable (😎).

Reply:
So the problem is that you don't want to have to query, say s3 to get the signed url. We had this issue in current xpub where we were doing that on autosave every 5 seconds, which made no sense (iirc) so you had to ask for it specifically.

Would the header be:
```
Prefer: signed-url
```
?

Reply:
Yeah, or something like `Prefer: public-link=signed` if these a variants.

Response needs a `Preference-Applied` and `Vary`.